### PR TITLE
Build: Add e2e-tests-dev task

### PR DESCRIPTION
**Pull request by yannbf**
## What I did

We have been running e2e tests against built Storybooks, but not against Storybooks in dev mode.
This PR solves that!

## How to test

Run a e2e-tests-dev for template, e.g. `yarn task --task e2e-tests-dev --start-from auto --template react-vite/default-ts`

## Checklist

- [x] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [ ] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [x] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [x] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

### CODE REVIEW

File: scripts/task.ts [hunk 1]
```diff
@@ -316,11 +318,16 @@ async function runTask(task: Task, details: TemplateDetails, optionValues: Passe
...
 
+const controllers: AbortController[] = [];
+
 async function run() {
   const allOptionValues = await getOptionsOrPrompt('yarn task', options);
 
   const { task: taskKey, startFrom, junit, ...optionValues } = allOptionValues;
 
+  // In case other tasks need to know
+  process.env.SELECTED_TASK = taskKey;
```
Comment (tmeasday) [inline scripts/task.ts:329]: Can we pass this as part of `TemplateDetails` rather than using a global?

File: scripts/tasks/dev.ts [hunk 2]
```diff
@@ -3,26 +3,32 @@ import detectFreePort from 'detect-port';
...
   dependsOn: ['sandbox'],
   async ready() {
-    return (await detectFreePort(START_PORT)) !== START_PORT;
+    return (await detectFreePort(PORT)) !== PORT;
   },
   async run({ sandboxDir, codeDir }, { dryRun, debug }) {
     const controller = new AbortController();
+    const devCommand = `yarn storybook --port ${PORT}${
+      process.env.SELECTED_TASK === 'dev' ? '' : ' --ci'
+    }`;
```
Comment (tmeasday) [inline scripts/tasks/dev.ts:21]: Maybe we could just add a flag for this to the task runner? IDK.

**Review by tmeasday — CHANGES_REQUESTED** [review]
Can we factor things so we share the code between the two files?

We could  have them read the same lib.

We could also consider a more complicated system where we can run any of the final tasks against either `dev` or `build` (arguably we could also run the `test-runner` task against `dev` also?) and have a flag that chooses between them, or a `e2e-tests-dev` task that depends on `dev` and `e2e-tests` somehow? Probably not the right time to think about that!

**Comment by tmeasday** [comment]
Also, let's make a ticket where we delete smoke testing from the repo if we all agree we don't need it any more.

File: scripts/task.ts (hunk 1 above)
Comment (yannbf) [reply scripts/task.ts:329]: We could, though the select task has nothing to do with the template, right

File: scripts/tasks/dev.ts (hunk 2 above)
Comment (yannbf) [reply scripts/tasks/dev.ts:21]: I thought of that, but didn't know if it was the right approach as it didn't seem to be related to TemplateDetails

**Comment by yannbf** [comment]
> Can we factor things so we share the code between the two files?
> 
> We could have them read the same lib.
>
> We could also consider a more complicated system where we can run any of the final tasks against either dev or build (arguably we could also run the test-runner task against dev also?) and have a flag that chooses between them, or a e2e-tests-dev task that depends on dev and e2e-tests somehow? Probably not the right time to think about that!

Would you like to take a stab at it @tmeasday? Feel free to change this PR or make a telescoping one!

**Review by shilman — COMMENTED** [review]
@tmeasday @yannbf Can we limit our dev coverage to 1 webpack-based project and 1 vite-based project and perhaps 1-2 others at most? If not, we're nearly doubling our CI costs, right? If we adopt `nx affected` and are able to dramatically reduce costs that way, we can look at testing dev more aggressively. WDYT?

File: scripts/tasks/dev.ts (hunk 2 above)
Comment (yannbf) [reply scripts/tasks/dev.ts:21]: Done!

**Comment by yannbf** [comment]
@tmeasday @shilman I worked on it based on your feedback. Please check!

**Review by tmeasday — APPROVED** [review]
LGTM. We should talk about a better way to disable a given task at a given cadence but this way is good for now!

File: .circleci/config.yml [hunk 3]
```diff
@@ -503,19 +548,28 @@ workflows:
...
+      - e2e-dev:
+          requires:
+            - create-sandboxes
+      - test-runner-production:
           requires:
             - build-sandboxes
-      - e2e-sandboxes:
+      - test-runner-dev:
           requires:
             - build-sandboxes
```
Comment (ndelangen) [inline .circleci/config.yml:574]: I think this should be `create-sandboxes` ?

Comment (yannbf) [reply .circleci/config.yml]: great catch!

**Comment by yannbf** [comment]
I'd love some help to understand why these tests fail in CI:

image

**Comment by tmeasday** [comment]
@yannbf if you are stuck I can have a go at figuring it out tomorrow perhaps?

**Comment by yannbf** [comment]
> @yannbf if you are stuck I can have a go at figuring it out tomorrow perhaps?

Thank you!

**Pull request merged**